content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,326 @@
+add(5, "foo") -- TypeError: Type `string` could not be converted into `number`.
+```
+
+To define a return type, put a `:` operator at the end of the function definition
+
+```lua
+local function add(x: number, y: number): number
+```
+
+To return multiple types, place all return types in brackets
+
+```lua
+local function FindSource(script: BaseScript, pattern: string): (string, number)
+```
+
+### Defining a Functional Type
+
+A functional type can be defined by using the syntax `(in) -> out`, for example, the type signatures of the functions above is
+
+```lua
+type add = (x: number, y: number) -> number
+type FindSource = (script: BaseScript, pattern: string) -> (string, number)
+```
+
+## Table Types
+
+Luau does not have a simple `table` type, instead, table types are defined using `{}` syntax. At the most basic level, a table can be defined with the syntax `{type}`
+
+```lua
+local numbers: {number} = {1, 2, 3, 4, 5}
+```
+
+Index types can be defined using `{[indexType] = valueType}`
+
+```lua
+local messagesFromPeople: {[string]: any} = {
+	Dave = "Hello",
+	Cat = "meow"
+}
+```
+
+Tables can also have explicit string indicies defined in a type
+
+```lua
+type Car = {
+	Speed: number,
+	Drive: (Car) -> ()
+}
+```
+
+## Variadics
+
+Lets assume a function that calculates the sum of a variadic ammount of numbers
+
+```lua
+local function addLotsOfNumbers(...)
+	local sum = 0
+
+	for _, v in {...} does
+		sum += v
+	end
+
+	return sum
+end
+```
+
+As expected, this can take any value and wont complain if you give it an invalid type such as a `string`.
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5)) -- 15
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- attempt to add string to number
+```
+
+Instead, we can assign a type to the `...` just like how you'd assign any other type
+
+```lua
+local function addLotsOfNumbers(...: number)
+```
+
+And now, the second line raises a type error
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5))
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- TypeError: Type `string` could not be converted into `number`.
+```
+
+However, with a type definition, the following statement does not work
+
+```lua
+type addLotsOfNumbers = (...: number) -> number -- Expected type, got ':'
+```
+
+Instead, use the syntax `...type` to define the variadic type
+
+```lua
+type addLotsOfNumbers = (...number) -> number
+```
+
+## Unions and Intersections
+It is possible to define a type to be two or more types using a union or intersection
+
+```lua
+type NumberOrString = number | string
+type Object1 = {foo: string}
+type Object2 = {bar: number}
+type Object1and2 = Object1 & Object2 -- {foo: string} & {bar: number}
+```
+
+## Defining an Inferred Type
+
+The `typeof` function can be used in a type definition to define an inferred type
+
+```lua
+type Car = typeof {
+	Speed = 0,
+	Wheels = 4
+} --> Car: {Speed: number, Wheels: number}
+```
+
+One interesting use is to define a metatable type using `setmetatable` inside `typeof`

```

Comment:
```suggestion
One interesting use is to define a metatable type using `setmetatable` inside `typeof`.
```

Casual phrasing. Consider "one way to use" or other equivalent phrasing if you think it's needed.